### PR TITLE
feat: add implementations of `CheckInit` and `OverrideInit`

### PR DESCRIPTION
**Pull request by AayushSabharwal**
@TorkelE @isaacsas this should make it so that your rebuild function is something along the lines of:

```julia
function rebuild(prob; u0, p)
   newprob = remake(prob; u0, p)
   u0, p, success = SciMLBase.get_initial_values(newprob, newprob, newprob.f, OverrideInit(), Val(SciMLBase.isinplace(newprob.f); nlsolve_alg = ...)
   @assert success
   newprob = remake(newprob; u0, p)
   return newprob
end
```

## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

**Comment by isaacsas** [comment]
OK, awesome! Thanks for sticking with this.

If you feel this is at a point where it would be helpful to test against Catalyst let me know. We have master capped right now to SciMLBase 2.57.1 as that was the last time tests passed (and we wanted to be able to work / merge other PRs).

File: Project.toml [hunk 1]
```diff
@@ -70,6 +70,7 @@ LinearAlgebra = "1.10"
 Logging = "1.10"
 Makie = "0.20, 0.21"
 Markdown = "1.10"
+NonlinearSolve = "2, 3"
```
Comment (ChrisRackauckas) [inline Project.toml:73]:
```suggestion
NonlinearSolve = "4"
```

File: src/initialization.jl [hunk 2]
```diff
@@ -30,3 +36,155 @@ struct OverrideInitData{IProb, UIProb, IProbMap, IProbPmap}
...
+    p = parameter_values(integrator)
+    t = current_time(integrator)
+    M = f.mass_matrix
+
+    algebraic_vars = [all(iszero, x) for x in eachcol(M)]
+    algebraic_eqs = [all(iszero, x) for x in eachrow(M)]
+    (iszero(algebraic_vars) || iszero(algebraic_eqs)) && return
+    update_coefficients!(M, u0, p, t)
+    tmp = _evaluate_f_ode(integrator, f, isinplace, u0, p, t)
+    tmp .= ArrayInterface.restructure(tmp, algebraic_eqs .* _vec(tmp))
```
Comment (ChrisRackauckas) [inline src/initialization.jl:115]: This allocates

File: src/initialization.jl [hunk 3]
```diff
@@ -30,3 +36,155 @@ struct OverrideInitData{IProb, UIProb, IProbMap, IProbPmap}
...
+"""
+function get_initial_values(prob::ODEProblem, integrator, f, alg::CheckInit,
+        isinplace::Union{Val{true}, Val{false}}; kwargs...)
+    u0 = state_values(integrator)
+    p = parameter_values(integrator)
+    t = current_time(integrator)
+    M = f.mass_matrix
+
+    algebraic_vars = [all(iszero, x) for x in eachcol(M)]
+    algebraic_eqs = [all(iszero, x) for x in eachrow(M)]
```
[the remark this answers is not in the document]
Comment (AayushSabharwal) [reply src/initialization.jl:111]: Cached what, exactly? 😅 I copied almost all of this from the existing implementation

File: src/initialization.jl (hunk 2 above)
Comment (AayushSabharwal) [reply src/initialization.jl:115]: Somehow changing this to `tmp = algebraic_eqs .* _vec(tmp)` allocates _more_

Comment (AayushSabharwal) [reply src/initialization.jl:115]: At least, according to `@allocated`/`@allocations`

**Pull request merged**